# Web Scraping Lab

You will find in this notebook some scrapy exercises to practise your scraping skills.

**Tips:**

- Check the response status code for each request to ensure you have obtained the intended contennt.
- Print the response text in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.

- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Beautiful Soup Doc](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [Urllib](https://docs.python.org/3/library/urllib.html#module-urllib)
- [re lib](https://docs.python.org/3/library/re.html)
- [lxml lib](https://lxml.de/)
- [Scrapy](https://scrapy.org/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### Below are the libraries and modules you may need. `requests`,  `BeautifulSoup` and `pandas` are imported for you. If you prefer to use additional libraries feel free to uncomment them.

In [1]:
import requests as req
from bs4 import BeautifulSoup
import pandas as pd
# from pprint import pprint
# from lxml import html
# from lxml.html import fromstring
# import urllib.request
# from urllib.request import urlopen
# import random as rd
# import re
# import scrapy

#### Download, parse (using BeautifulSoup), and print the content from the Trending Developers page from GitHub:

In [2]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [3]:
#your code

html = req.get(url).content

html[:100]

b'\n\n<!DOCTYPE html>\n<html lang="en" data-color-mode="auto" data-light-theme="light" data-dark-theme="d'

In [4]:
soup = BeautifulSoup(html)

In [5]:
# no lo he impreso por vuestra salud mental, correctores 💕
# print(soup)

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names. You can achieve this using Chrome DevTools.

1. Use BeautifulSoup to extract all the html elements that contain the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below:

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [6]:
len(soup.find_all('article'))

48

In [14]:
author = soup.find_all('article')

# author[0]

In [15]:
author[2].find_all('a')[2].text.strip()

'Alex Stokes'

In [16]:
authors = []

for i in range(len(soup.find_all('article'))):
    try:
        author = soup.find_all('article')[i].find_all('a')[2].text.strip()
        authors.append(author)
        
    except:
        pass
    
authors

['Ismail Pelaseyed',
 'Alex Stokes',
 'Don Jayamanne',
 'Stefan Prodan',
 'Travis Cline',
 'Oliver',
 'dgtlmoon',
 'Guillaume Klein',
 'lllyasviel',
 'Angelos Chalaris',
 'Jon Skeet',
 'Chris Banes',
 'Carlos Scheidegger',
 'Anders Eknert',
 'Emil Ernerfeldt',
 'Andrea Aime',
 'Mattt',
 'Meng Zhang',
 'Shahed Nasser',
 'dkhamsing',
 'Kieron Quinn',
 'afc163',
 'Alan Donovan',
 'Lee Calcote',
 'Costa Huang']

#### Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [17]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/python?since=daily'

In [18]:
#your code
html = req.get(url).content

soup = BeautifulSoup(html)

In [21]:
repository = soup.find_all('article')

# repository

In [22]:
repository[0].find_all('a')[1].text.strip().split()[-1]

'ChatGLM3'

In [23]:
repositories = []

for i in range(len(soup.find_all('article'))):
    try:
        repository = soup.find_all('article')[i].find_all('a')[1].text.strip().split()[-1]
        repositories.append(repository)
    except:
        pass
    
set(repositories)

{'ChatDev',
 'ChatGLM3',
 'CogVLM',
 'Fooocus',
 'PaddleOCR',
 'Star',
 'Wonder3D',
 'autotrain-advanced',
 'bisheng',
 'discord_bot',
 'face_recognition',
 'haystack',
 'langflow',
 'nnUNet',
 'poetry',
 'public-apis',
 'rag-demystified',
 'raven',
 'reflex',
 'system-design-primer',
 'ungoogled-chromium',
 'whisper'}

#### Display all the image links from Walt Disney wikipedia page

In [24]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [25]:
#your code
html = req.get(url).content

soup = BeautifulSoup(html)

In [26]:
list_soups = soup.find('div', {'id': 'bodyContent'}).find_all('img')[2:]

Le recorté las dos primeras imagenes porque no tienen nada que ver con Walt Disney

In [27]:
list_soups[0]['src']

'//upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG'

In [28]:
images = []

for i in range(len(list_soups)):
    images.append(list_soups[i]['src'])
    
images[:5]

['//upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG',
 '//upload.wikimedia.org/wikipedia/commons/thumb/8/87/Walt_Disney_1942_signature.svg/150px-Walt_Disney_1942_signature.svg.png',
 '//upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Walt_Disney_Birthplace_Exterior_Hermosa_Chicago_Illinois.jpg/220px-Walt_Disney_Birthplace_Exterior_Hermosa_Chicago_Illinois.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Walt_Disney_envelope_ca._1921.jpg/220px-Walt_Disney_envelope_ca._1921.jpg',
 '//upload.wikimedia.org/wikipedia/en/thumb/4/4e/Steamboat-willie.jpg/220px-Steamboat-willie.jpg']

#### Retrieve an arbitary Wikipedia page of "Python" and create a list of links on that page

In [4]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 

In [5]:
#your code
html = req.get(url).content

soup = BeautifulSoup(html, "html.parser")

In [6]:
ending = soup.find_all('a', {'title': 'Monty Python'})[0]['href']

ending

'/wiki/Monty_Python'

In [7]:
mpython = f'https://en.wikipedia.org{ending}'

In [8]:
mpython

'https://en.wikipedia.org/wiki/Monty_Python'

In [9]:
soup = BeautifulSoup(req.get(mpython).content)

In [10]:
links = soup.find('div', {'id': 'bodyContent'}).find_all('a')
links[:10]

[<a href="/wiki/Monty_Python%27s_Flying_Circus" title="Monty Python's Flying Circus">Monty Python's Flying Circus</a>,
 <a href="/wiki/Pythonesque_(play)" title="Pythonesque (play)">Pythonesque (play)</a>,
 <a href="/wiki/The_Pythons_(film)" title="The Pythons (film)">The Pythons (film)</a>,
 <a class="mw-file-description" href="/wiki/File:Flyingcircus_2.jpg"><img class="mw-file-element" data-file-height="273" data-file-width="365" decoding="async" height="187" src="//upload.wikimedia.org/wikipedia/en/thumb/c/cb/Flyingcircus_2.jpg/250px-Flyingcircus_2.jpg" srcset="//upload.wikimedia.org/wikipedia/en/c/cb/Flyingcircus_2.jpg 1.5x" width="250"/></a>,
 <a href="#cite_note-1">[1]</a>,
 <a href="/wiki/Comedic_genres" title="Comedic genres">Genres</a>,
 <a href="/wiki/Satire" title="Satire">Satire</a>,
 <a href="/wiki/Surreal_humour" title="Surreal humour">surreal humour</a>,
 <a href="/wiki/Black_comedy" title="Black comedy">black comedy</a>,
 <a class="mw-redirect" href="/wiki/Wordplay" tit

In [13]:
link_list = []

for link in links:
        href = link.get("href")
        if href and "/wiki/" in href:
            link_list.append(href) # podría hacer append a f'https://en.wikipedia.org{href}' pero no sé si eso es lo que se pide
        
link_list[:10]

['/wiki/Monty_Python%27s_Flying_Circus',
 '/wiki/Pythonesque_(play)',
 '/wiki/The_Pythons_(film)',
 '/wiki/File:Flyingcircus_2.jpg',
 '/wiki/Comedic_genres',
 '/wiki/Satire',
 '/wiki/Surreal_humour',
 '/wiki/Black_comedy',
 '/wiki/Wordplay',
 '/wiki/Wit']

#### Number of Titles that have changed in the United States Code since its last release point 

In [37]:
# This is the url you will scrape in this exercise
url = 'http://uscode.house.gov/download/download.shtml'

In [38]:
html = req.get(url).content

soup = BeautifulSoup(html, features='xml')

In [39]:
changed = soup.find_all('div', {'class': 'usctitlechanged'})

changed

[<div class="usctitlechanged" id="us/usc/t20">
 
           Title 20 - Education
 
         </div>,
 <div class="usctitlechanged" id="us/usc/t38">
 
           Title 38 - Veterans' Benefits <span class="footnote"><a class="fn" href="#fn">٭</a></span>
 </div>]

In [40]:
changes = []

for i in changed:
    changes.append(i.text.strip())
    
changes

['Title 20 - Education', "Title 38 - Veterans' Benefits ٭"]

#### A Python list with the top ten FBI's Most Wanted names 

In [41]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'

In [42]:
html = req.get(url).content

soup = BeautifulSoup(html)

Intentando scrapear esta web me devuelve siempre mi ip, me da que el fbi no quiere que le toqueteen la web.

####  20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [31]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake_information/'

In [32]:
#your code
html = req.get(url).content

soup = BeautifulSoup(html)

In [54]:
table = soup.find_all('table')[2]

In [55]:
head = table.find_all('th')

In [58]:
cols = [e.text.replace('\n', ' ') for e in head if e.text][1:]

In [59]:
cols

['Date & TimeUTC',
 'Lat.degrees',
 'Lon.degrees',
 'Depthkm',
 'Mag.[+]',
 'Region']

In [63]:
table.find_all('tr') # la tabla no carga hasta que se scrollea, por lo que no se puede con bs4

[<tr><th class="thico"></th><th class="citiz" colspan="2"><div>Citizen<br/>response</div><div><div class="dm comm"></div><div class="dm pic"></div></div></th>
 <th class="tbdat">Date &amp; Time<div>UTC</div></th><th class="tblat">Lat.<div>degrees</div></th><th class="tblon">Lon.<div>degrees</div></th><th class="tbdep">Depth<div>km</div></th><th class="tbmag">Mag.<span>[+]</span></th><th class="tbreg">Region</th></tr>]

#### Display the date, days, title, city, country of next 25 hackathon events as a Pandas dataframe table

In [24]:
# This is the url you will scrape in this exercise
url ='https://hackevents.co/hackathons'

In [25]:
#your code

#### Count number of tweets by a given Twitter account.

You will need to include a ***try/except block*** for account names not found. 
<br>***Hint:*** the program should count the number of tweets for any provided account

In [64]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

#### Number of followers of a given twitter account

You will need to include a ***try/except block*** in case account/s name not found. 
<br>***Hint:*** the program should count the followers for any provided account

In [65]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/chocu_'

In [ ]:
html = req.get(url).content

soup = BeautifulSoup(html)

In [29]:
soup.find('div', {'class': 'css-1dbjc4n'}).text # it always retrieve that message

#### List all language names and number of related articles in the order they appear in wikipedia.org

In [16]:
# This is the url you will scrape in this exercise
url = 'https://www.wikipedia.org/'

In [17]:
html = req.get(url).content

soup = BeautifulSoup(html, 'html.parser')

In [18]:
box = soup.find('div', {'class': "central-featured"})

# box

In [19]:
langs = box.find_all('strong')

langs

[<strong>English</strong>,
 <strong>日本語</strong>,
 <strong>Español</strong>,
 <strong>Русский</strong>,
 <strong>Deutsch</strong>,
 <strong>Français</strong>,
 <strong>Italiano</strong>,
 <strong>中文</strong>,
 <strong>Português</strong>,
 <strong><bdi dir="rtl">العربية</bdi></strong>]

In [20]:
nums = box.find_all('small')

nums

[<small><bdi dir="ltr">6 715 000+</bdi> <span>articles</span></small>,
 <small><bdi dir="ltr">1 387 000+</bdi> <span>記事</span></small>,
 <small><bdi dir="ltr">1 892 000+</bdi> <span>artículos</span></small>,
 <small><bdi dir="ltr">1 938 000+</bdi> <span>статей</span></small>,
 <small><bdi dir="ltr">2 836 000+</bdi> <span>Artikel</span></small>,
 <small><bdi dir="ltr">2 553 000+</bdi> <span>articles</span></small>,
 <small><bdi dir="ltr">1 826 000+</bdi> <span>voci</span></small>,
 <small><bdi dir="ltr">1 377 000+</bdi> <span class="jscnconv" data-hans="条目" data-hant="條目">条目 / 條目</span></small>,
 <small><bdi dir="ltr">1 109 000+</bdi> <span>artigos</span></small>,
 <small><bdi dir="ltr">1 217 000+</bdi> <span>مقالة</span></small>]

In [21]:
len(nums), len(langs)

(10, 10)

In [22]:
nums = [e.text.replace('\xa0', '').split('+')[0] for e in nums]

langs = [e.text for e in langs]

In [23]:
dict(zip(langs, nums))

{'English': '6715000',
 '日本語': '1387000',
 'Español': '1892000',
 'Русский': '1938000',
 'Deutsch': '2836000',
 'Français': '2553000',
 'Italiano': '1826000',
 '中文': '1377000',
 'Português': '1109000',
 'العربية': '1217000'}

#### A list with the different kind of datasets available in data.gov.uk 

In [78]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'

In [79]:
html = req.get(url).content

soup = BeautifulSoup(html, 'html.parser')

In [87]:
soup.find('ul', {'class': 'govuk-list dgu-topics__list'}).find_all('li')[0].text.split('\n')[0]

''

In [90]:
ds_list = soup.find('ul', {'class': 'govuk-list dgu-topics__list'}).find_all('li')

ds = [e.text.split('\n')[1] for e in ds_list]

ds

['Business and economy',
 'Crime and justice',
 'Defence',
 'Education',
 'Environment',
 'Government',
 'Government spending',
 'Health',
 'Mapping',
 'Society',
 'Towns and cities',
 'Transport',
 'Digital service performance',
 'Government reference data']

#### Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [34]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'

In [35]:
#your code

### BONUS QUESTIONS

#### Scrape a certain number of tweets of a given Twitter account.

In [36]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [37]:
# your code

#### IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [38]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

In [39]:
# your code

#### Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [24]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [28]:
html = req.get(url).text

soup = BeautifulSoup(html, 'html.parser')

In [30]:
soup # no se puede

<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
</body>
</html>

#### Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [42]:
#https://openweathermap.org/current
city = city=input('Enter the city:')
url = 'http://api.openweathermap.org/data/2.5/weather?'+'q='+city+'&APPID=b35975e18dc93725acb092f7272cc6b8&units=metric'

Enter the city:madrid


In [43]:
# your code

#### Book name,price and stock availability as a pandas dataframe.

In [44]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

In [45]:
#your code